drivers: crc: Add hardware-based CRC driver

zkat2 opened the pull request:
Adds hardware support for CRC-8-CCITT and CRC-32-IEEE calculations and introduces a devicetree binding for STM32 CRC.

Commit:
drivers: crc: Add hardware-based CRC driver

Adds support for hardware-based  CRC-8-CCITT and CRC-32-IEEE calculations.

Signed-off-by: Zoe Kaute <zoe.kaute@brillpower.com>

erwango commented:
@mathieuchopstm Would you mind having a look ?

mathieuchopstm commented on drivers/crc/crc_hw.h:
```diff
@@ -0,0 +1,253 @@
+	 * else return the setup function which will return 0
+	 */
+#ifdef CONFIG_CRC_THREADSAFE
+	const struct crc_data *data = dev->data;
```
AFAIK, `device->data` is driver-private and opaque to Zephyr. I don't think using it for this purpose is OK.

mathieuchopstm commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,193 @@
+
+	/* Compute the CRC of Data Buffer array */
+	for (index = 0; index < (len / 4); index++) {
+		data = (uint32_t)((num_data[4 * index + 3] << 24) | (num_data[4 * index + 2] << 16)
+							(num_data[4 * index + 1] << 8) | num_data[4 * index]);
```
I think `sys_get_be32` from `<zephyr/sys/byteorder.h>` should be used instead.
```suggestion
		data = sys_get_be32(num_data);
```

mathieuchopstm commented on drivers/crc/crc_hw.h:
```diff
@@ -0,0 +1,253 @@
+ * @struct crc_driver_api
+ * @brief Publicly available API accessible as part of device struct
+ */
+struct crc_driver_api {
```
This structure declaration is lacking the `__subsystem` annotation (from `<zephyr/toolchain.h>`).

mathieuchopstm commented on drivers/crc/crc_hw.h:
```diff
@@ -0,0 +1,253 @@
+ * @struct crc_driver_api
+ * @brief Publicly available API accessible as part of device struct
+ */
+struct crc_driver_api {
+	crc_api_setup_crc_8_ccitt setup_crc_8_ccitt;
+	crc_api_crc_8_ccitt crc_8_ccitt;
+	crc_api_verify_crc_8_ccitt verify_crc_8_ccitt;
+	crc_api_setup_crc_32_ieee setup_crc_32_ieee;
+	crc_api_crc_32_ieee crc_32_ieee;
+	crc_api_verify_crc_32_ieee verify_crc_32_ieee;
+};
```
+1

I think an API inspired by the [cryptography subsystem](https://github.com/zephyrproject-rtos/zephyr/blob/7f8cc43a0b5e675569866b7364ab34ba8ebd4e0a/include/zephyr/crypto/crypto.h#L67-L91) could be a good fit.

Looking at the API exposed by STM32 LL, a starting point could be:
```c
enum crc_size {
   CRC_7,
   CRC_8,
   CRC_16,
   CRC_32,
   CRC_64
};

typedef crc_polynomial_t uint64_t;
typedef crc_result_t uint64_t;

struct crc_ctx {
   /* we could imagine the CRC state being
      stored here to allow concurrent
      operations?
    */
   /* result could also possibly be stored here */
};

#define CRC_FLAG_REVERSE_INPUT BIT(0)
#define CRC_FLAG_REVERSE_OUTPUT BIT(1)

__subsystem struct crypto_driver_api {
    int (*crc_begin)(const struct device *dev, struct crc_ctx *ctx, enum crc_mode mode, uint32_t flags, crc_polynomial_t polynomial);
    int (*crc_update)(const struct device *dev, struct crc_ctx *ctx, const void *buffer, size_t bufsize);
    /* semantics to define - does it behave as if updating byte-by-byte? crc-size by crc-size? */
    int (*crc_finalize)(const struct device *dev, struct crc_ctx *ctx, crc_result_t *result);
};
```

mathieuchopstm commented on drivers/crc/crc_hw.h:
```diff
@@ -0,0 +1,253 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
```
The Doxygen API documentation comment is missing
```suggestion
/**
 * @brief CRC driver APIs
 * @defgroup crc_interface CRC driver APIs
 * @ingroup io_interfaces
 * @{
 */
```

mathieuchopstm commented on drivers/crc/crc_hw.h:
```diff
@@ -0,0 +1,253 @@
+#endif /* CONFIG_CRC_THREADSAFE */
+};
+
+#ifdef CONFIG_CRC_HW
```
I have not seen such conditional exclusion in other driver API headers, and it appears to go against [the API Guidelines](https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#conditional-data-and-apis).

mathieuchopstm commented on drivers/crc/crc_hw.h:
```diff
@@ -0,0 +1,253 @@
+}
+
+#endif /* CONFIG_CRC_HW */
+
```
The Doxygen group ending comment is missing
```suggestion
/**
 * @}
 */
```

mathieuchopstm commented on drivers/crc/crc_hw.h:
```diff
@@ -0,0 +1,253 @@
+	return api->verify_crc_32_ieee(dev, expected, buf, len);
+}
+
+#endif /* CONFIG_CRC_HW */
```
(see comment on matching `#if`)

mathieuchopstm commented on drivers/crc/crc_hw.h:
This file should reside in `include/zephyr/drivers` and be named `crc.h`.

mathieuchopstm commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,193 @@
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/init.h>
+#include "stm32_ll_crc.h"
+#include "crc_hw.h"
```
Should be changed to `<zephyr/drivers/crc.h>`.

mathieuchopstm reviewed (commented):
Same architectural concerns as @pdgendt about the API.
Ignoring this, code looks mostly OK, with a few modifications to make here and there, to match other driver subsystems.

mathieuchopstm commented on drivers/crc/Kconfig:
```diff
@@ -0,0 +1,30 @@
+
+source "drivers/crc/Kconfig.stm32"
+
+endif # CRC
```
```suggestion
endif # CRC_HW
```

mathieuchopstm commented on drivers/crc/Kconfig:
```diff
@@ -0,0 +1,30 @@
+config CRC_INIT_PRIORITY
+	int "CRC init priority"
+	default KERNEL_INIT_PRIORITY_DEVICE
+	depends on CRC_HW
```
Unnecessary (already protected by `if CRC_HW`).

mathieuchopstm commented on drivers/crc/Kconfig:
```diff
@@ -0,0 +1,30 @@
+config CRC_THREADSAFE
+	bool "Make HW CRC threadsafe"
+	default n
+	depends on CRC_HW
```
Unnecessary.

mathieuchopstm commented on drivers/crc/Kconfig.stm32:
```diff
@@ -0,0 +1,10 @@
+config CRC_STM32
+	bool "STM32 CRC driver"
+	default n
+	depends on DT_HAS_ST_STM32_CRC_ENABLED && CRC_HW
```
```suggestion
	depends on DT_HAS_ST_STM32_CRC_ENABLED
```
(comes implicitly from `Kconfig` which `source`s this - see https://docs.zephyrproject.org/latest/build/kconfig/tips.html#lack-of-conditional-includes)

pdgendt commented on drivers/crc/Kconfig.stm32:
```diff
@@ -0,0 +1,10 @@
+
+config CRC_STM32
+	bool "STM32 CRC driver"
+	default n
```
Enable this by default as there is a devicetree check dependency.

```suggestion
	default y
```

pdgendt commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,191 @@
+ *		is used within multiple threads
+ */
+struct crc_data {
+	struct k_mutex mutex;
```
Why have this for all drivers? This should be left up to the driver if concurrent calls need to be safeguarded IMO.

pdgendt commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,232 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+#include <errno.h>
+#include <soc.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/crc.h>
+#include "stm32_ll_crc.h"
+
+LOG_MODULE_REGISTER(crc_stm32);
```
You need to pass the logging level from Kconfig.

Try to group includes and have some order like `system` > `zephyr` > `SoC`

```c
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(crc_stm32, CONFIG_CRC_HW_LOG_LEVEL);

#include <errno.h>

#include <zephyr/device.h>
#include <zephyr/drivers/clock_control/stm32_clock_control.h>
#include <zephyr/drivers/crc.h>

#include <soc.h>
#include <stm32_ll_crc.h>
```

pdgendt reviewed (commented):
This is going in the right direction IMO, there are some compliance issues that need attention.

erwango commented on drivers/crc/Kconfig.stm32:
```diff
@@ -0,0 +1,10 @@
+	bool "STM32 CRC driver"
+	default n
+	depends on DT_HAS_ST_STM32_CRC_ENABLED
+	select USE_STM32_LL_CRC
```
`USE_STM32_LL_F00` are required only when making use of c LL files, in your case unless you use `LL_CRC_DeInit` function, inlcuding `stm32_ll_crc.h` is sufficient to use the API.

mathieuchopstm commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,191 @@
+	crc_result_t result;         /**< Store the CRC result */
+};
+
+typedef int (*crc_api_lock)(const struct device *dev, k_timeout_t timeout);
+typedef int (*crc_api_unlock)(const struct device *dev);
```
These are out-of-scope imho: If driver wants to lock device when in use (i.e., `begin` called but `finish`/`verify`/whatever not called), it should be allowed to do so, but not forced to.

erwango commented on dts/bindings/crc/st,stm32-crc.yaml:
```diff
@@ -0,0 +1,25 @@
+    required: true
+    description: Clock information
+
+  timeout:
+    type: int
+    default: 0
+    description: |
+      Timeout period (milliseconds) for use when acquiring mutex lock for 8-bit CRC
+      in thread-safe mode.
+      Min=0, Max=65535, >10000=K_FOREVER
```
As this is pure SW behavior, it should rather be configured by Kconfig.
https://docs.zephyrproject.org/latest/build/dts/dt-vs-kconfig.html
> In other words, devicetree mainly deals with hardware, and Kconfig with software.

A reason to use device tree could that you want to be able to configure different values per instance, but this doesn't look the case here.

mathieuchopstm commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,191 @@
+			     const void *initial_value);
+typedef int (*crc_api_update)(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+			      size_t bufsize);
+typedef int (*crc_api_verify)(const struct device *dev, struct crc_ctx *ctx, crc_result_t expected);
```
Why not something like:
```suggestion
typedef int (*crc_api_finish)(const struct device *dev, struct crc_ctx *ctx, crc_result_t *result);
```
Current API unnecessarily restricts use cases. `verify` can be trivially implemented in-header:
```c
int crc_finish_and_verify(const struct device *dev, struct crc_ctx *ctx, crc_result_t expected)
{
   crc_result_t crc_res;
   int res = crc_finish(dev, ctx, &res);
   if (res < 0) {
      return res;
   }

   if (crc_res != expected) {
      return -EINVAL; //or w/e
   }
   return 0;
}
```

zkat2 replied:
Would you be able to clarify the intended use of the `result` parameter in the proposed `crc_api_finish` function? Is the goal is to provide direct access to the CRC result without requiring users to manipulate the ctx structure directly, or something else?

mathieuchopstm replied:
Yes, exactly. Currently, the `struct crc_ctx` contains the `result` after each `update`, but it makes more sense to only compute `result` once, after all bytes to checksum have been provided via `update`s. This is just a proposal.
In any case, current implementation of `crc_api_verify` in driver should *not* be in the driver because it has nothing hardware-specific.

zkat2 replied:
In the current implementation,  the crc_update function handles the full data buffer, processes it, and updates the result in the ctx struct. I not sure I see the functionality of moving the result retrieval to the finish function

mathieuchopstm replied:
maybe it would not be necessary to return CRC result in `finish`, but in any case, if we want to follow the crypto API, we need a `begin` and `finish`

mathieuchopstm commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,191 @@
+struct crc_ctx {
+	enum crc_type type;          /**< Type of CRC algorithm */
+	enum crc_state state;        /**< State of CRC calculation */
+	const void *initial_value;   /**< Initial value for CRC calculation */
```
Why `const void *` type and not an integer type?

mathieuchopstm commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,191 @@
+			    uint32_t flags, crc_polynomial_t polynomial, const void *initial_value)
+{
+	const struct crc_driver_api *api = (const struct crc_driver_api *)dev->api;
+	ctx->type = type;
+	ctx->flags = flags;
+	ctx->polynomial = polynomial;
+	ctx->initial_value = initial_value;
+	ctx->state = CRC_STATE_IN_PROGRESS;
```
What if `ctx->state == CRC_STATE_IN_PROGRESS`?

Also, in this case, it would be wiser to have user fill up the structure, and only accept a `struct crc_ctx *ctx` as parameter, since it already contains all data.

erwango reviewed (changes requested):
Looks promising, thanks.

Some additional requests as you're introducing a brand new API:
- API doc, see doc/hardware/peripherals/dac.rst as a minimum doc page
- Add yourself as maintainer in MAINTAINERS.yaml
- Last (maybe the most important): please provide a sample or a test (both would be ideal ;)) in order:
  - to demonstrate how to use the API and how to configure a device
  - ensure the API and related drivers are compiled systematically in CI

mathieuchopstm commented on dts/bindings/crc/st,stm32-crc.yaml:
```diff
@@ -0,0 +1,25 @@
+    type: int
+    default: 0
+    description: |
+      Timeout period (milliseconds) for use when acquiring mutex lock for 8-bit CRC
+      in thread-safe mode.
```
```suggestion
      Timeout period (milliseconds) for use when locking CRC instance.
```

mathieuchopstm commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,232 @@
+
+		/* Last bytes specific handling */
+		if ((bufsize & 3) != 0) {
+			if ((bufsize & 3) == 1) {
+				LL_CRC_FeedData8(cfg->base, num_data[4 * index]);
+			}
+			if ((bufsize & 3) == 2) {
+				LL_CRC_FeedData16(cfg->base,
+						  (uint16_t)((num_data[4 * index + 1] << 8) |
+							     num_data[4 * index]));
+			}
+			if ((bufsize & 3) == 3) {
+				LL_CRC_FeedData16(cfg->base,
+						  (uint16_t)((num_data[4 * index + 1] << 8) |
+							     num_data[4 * index]));
+				LL_CRC_FeedData8(cfg->base, num_data[4 * index + 2]);
+			}
```
Not correct. This feeds data twice if `(bufsize & 3) == 3`.
Also, `sys_get_beXX` from `<zephyr/sys/byteorder.h>` should be used instead of manually unpacking.
```suggestion
			if ((bufsize & 3) == 3) {
			       LL_CRC_FeedData16(cfg->base, sys_get_be16(num_data + 4 * index));
			       LL_CRC_FeedData8(cfg->base, num_data[4 * index + 2]);
			} else if ((bufsize & 3) == 2) {
			       LL_CRC_FeedData16(cfg->base, sys_get_be16(num_data + 4 * index));
			} else { /* ((bufsize & 3) == 1) */
	                       LL_CRC_FeedData8(cfg->base, num_data[4 * index]);		
			}
```

mathieuchopstm commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,232 @@
+			}
+		}
+		uint32_t result = (~LL_CRC_ReadData32(cfg->base));
+		(ctx->result) = (crc_result_t)result;
```
Nit:
```suggestion
		ctx->result = (crc_result_t)result;
```

mathieuchopstm commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,232 @@
+
+		/* Compute the CRC of Data Buffer array*/
+		for (index = 0; index < (bufsize / 4); index++) {
+			data = (uint32_t)((num_data[4 * index + 3] << 24) |
+					  (num_data[4 * index + 2] << 16) |
+					  (num_data[4 * index + 1] << 8) | num_data[4 * index]);
```
Use `sys_get_be32`:
```suggestion
			data = sys_get_be32(num_data + 4 * index);
```

mathieuchopstm commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	const struct crc_cfg *cfg = dev->config;
+	const uint8_t *num_data = buffer;
```
Why `num_data`? This makes code hard to understand imo.

mathieuchopstm commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,232 @@
+			LL_CRC_FeedData8(cfg->base, num_data[index]);
+		}
+		uint8_t result = LL_CRC_ReadData8(cfg->base);
+		(ctx->result) = (crc_result_t)result;
```
Nit:
```suggestion
		ctx->result = (crc_result_t)result;
```

bp-amitchone commented on tests/drivers/crc/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+  crc.test:
+    integration_platforms:
+      - nucleo_g474re
+    platform_allow: nucleo_g474re modbuckboost_v2
```
`modbuckboost_v2` is an out-of-tree target so shouldn't be included here

```suggestion
    platform_allow: nucleo_g474re
```

bp-amitchone commented on tests/drivers/crc/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+      - nucleo_g474re
+    platform_allow: nucleo_g474re modbuckboost_v2
+    platform_exclude: native_posix
+    extra_args: EXTRA_CFLAGS=-Werror
+      DTC_OVERLAY_FILE="boards/nucleo_g474re.overlay"
```
```suggestion
    extra_args: "DTC_OVERLAY_FILE=boards/nucleo_g474re.overlay"
```

pdgendt commented on MAINTAINERS.yml:
```diff
@@ -1100,6 +1100,22 @@ Release Notes:
+"Drivers: CRC":
+  status: maintained
+  maintainers:
+    - zkat
```
Your username is `zkat2`, no?

pdgendt commented on drivers/crc/Kconfig:
Use tabs for indentation (and 2 spaces for the help description)

pdgendt replied:
Still not properly formatted.

pdgendt commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,232 @@
+
+#define DT_DRV_COMPAT st_stm32_crc
+
+struct crc_stm32_cfg
+{
```
`struct` braces need to follow the name.

```suggestion
struct crc_stm32_cfg {
```

pdgendt commented on drivers/crc/crc_stm32.c:
Not sure why compliance checks aren't failing, but as this is a new file, you should run `clang-format` on it before submitting.

pdgendt commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,163 @@
+
+	ctx->state = CRC_STATE_IN_PROGRESS;
+	return api->crc_begin(dev, ctx);
+}
```
You should add `syscalls`, see https://docs.zephyrproject.org/latest/kernel/usermode/syscalls.html

pdgendt commented on samples/drivers/crc/boards/nucleo_g474re.overlay:
```diff
@@ -0,0 +1,11 @@
+/ {
+	soc {
+		crc: crc@40023000 {
+			compatible = "st,stm32-crc";
```
As this is part of the SoC, this should be added there, not in a sample.

pdgendt commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,228 @@
+	struct k_mutex crc_mtx = data->mutex;
+	int res = 0;
+
+	/**
+	 * The lock count can be greater than 1 if the same thread acquires
+	 * a lock multiple times, so we have to iterate until lock_count
+	 * is zero
+	 */
+	while (crc_mtx.lock_count != 0) {
+		res = k_mutex_unlock(&data->mutex);
+
+		if (res != 0) {
+			return res;
+		}
+	}
```
You don't want to do this, the caller has to make sure to lock/unlock properly.

This also has a race condition if an unlock happens between line 50 & 51 in a higher priority thread/interrupt.

pdgendt commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,228 @@
+	}
+
+	default:
+		crc_unlock(dev);
+		return -ENOTSUP;
```
This locking mechanism looks really flawed, a mutex should be used to protect different API calls, you need another semaphore or something to allow for asynchronous behavior.

zkat2 replied:
What is the use case you are envisioning? The CRC peripheral has a single 32-bit register.

pdgendt commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,168 @@
+		return -EBUSY;
+	}
+
+	ctx->state = CRC_STATE_IN_PROGRESS;
```
Handling these states in the API itself feels like it wouldn't handle multithreaded applications very well.

pdgendt commented:
The project is currently in a feature freeze for LTS version 3.7, after the release we can continue with new features such as this PR.

mathieuchopstm reviewed (changes requested):
A few things to change:

WRT to the locking and timeout stuff, various remarks based on research in the tree:
* It would be preferable to use `k_sem` instead of `k_mutex`
  * semaphores are much cheaper to use
  * while they do not provide recursive acquisition, it doesn't appear needed here
* As far as I can tell, timeouts are mostly used to prevent application lockup due to external hardware conditions
  * e.g., communication failure with an SPI sensor
    * Example 1: [Various](https://github.com/zephyrproject-rtos/zephyr/blob/4ddd3ebd51fb85390c34bc580487f43d1144cd24/include/zephyr/drivers/can.h#L459-L462) [CAN](https://github.com/zephyrproject-rtos/zephyr/blob/4ddd3ebd51fb85390c34bc580487f43d1144cd24/include/zephyr/drivers/can.h#L483) API functions accept a `timeout` argument
    * Example 2: [`imx_usdhc.c` driver](https://github.com/zephyrproject-rtos/zephyr/blob/4ddd3ebd51fb85390c34bc580487f43d1144cd24/drivers/sdhc/imx_usdhc.c#L720) uses the `timeout` field of [`sdhc_command`](https://github.com/zephyrproject-rtos/zephyr/blob/4ddd3ebd51fb85390c34bc580487f43d1144cd24/include/zephyr/drivers/sdhc.h#L52)
    * Example 3: [`spi_context_wait_for_completion`](https://github.com/zephyrproject-rtos/zephyr/blob/4ddd3ebd51fb85390c34bc580487f43d1144cd24/drivers/spi/spi_context.h#L156-L164) in master mode will time out if transfer takes too long to complete
    * Example 4: [`MSPI_COMPLETION_TIMEOUT_TOLERANCE`](https://github.com/zephyrproject-rtos/zephyr/blob/4ddd3ebd51fb85390c34bc580487f43d1144cd24/drivers/mspi/Kconfig#L34-L38) Kconfig option, which is the closest to this PR
  * Unlike these examples, CRC HW accelerators should be standalone, so these lockups cannot happen

I think the following directions would be preferable to Kconfig-provided timeout:
* Change the `crc_begin` function to accept a `timeout`
  * More flexible than fixed thing in Kconfig
  * Avoids the `if (timeout > 10000) timeout = K_FOREVER;` mess
* Remove all `timeout` logic and make drivers fully blocking
  * This sounds to me like the most appropriate solution, because:
    * If an application _wants_ mutual exclusion with timeouts, it can always be implemented by it on top of the API
      * e.g., `struct my_crc { k_sem lock; const struct device *hw; }`
    * I fail to see a use case in which the application would be OK with skipping the CRC calculation _sometimes_ because of contention


As a last note, since this PR introduces a new API, it will have to undergo review from the Architecture Working Group, as described [here](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#peripheral-apis-hardware-related).

zkat2 commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,161 @@
+	CRC4,
+	CRC4_TI,
+	CRC7_BE,
+	CRC8,
+	CRC8_CCITT,
```
This enum was copied from include/zephyr/sys/crc.h for consistency with the existing software implementation

mathieuchopstm replied:
Good catch. It does look like the algorithms are different after checking the software implementation, so disregard this comment :sweat_smile:

zkat2 commented on include/zephyr/drivers/crc.h:
```diff
@@ -0,0 +1,161 @@
+	crc_result_t result;		/**< Store the CRC result */
+};
+
+typedef int (*crc_api_begin)(const struct device *dev, struct crc_ctx *ctx);
+typedef int (*crc_api_finish)(const struct device *dev, struct crc_ctx *ctx, const void *buffer,
+			      size_t bufsize);
```
Discontinuous buffers can be processed by calling the finish function multiple times to process each buffer and retrieving the final result from ctx once everything has been processed. I recognize that in this scenario it wouldn't make much sense to call finish multiple times so maybe this function can be renamed to update and the finish function can be omitted all together. The final result would be able to be retrieved directly from the ctx once all updates are completed. 

Is there a particular use case you had in mind for hardware peripheral that uses discontinuous buffers?

mathieuchopstm replied:
It is not possible to call `finish` multiple times in the current implementation:
https://github.com/zephyrproject-rtos/zephyr/blob/fac51a0f02e1cfb2dc7c122a98e813e982ef2e37/drivers/crc/crc_stm32.c#L103-L109

If several consecutive calls of `finish` is what you had in mind, then maybe it would be a better fit to have a single `crc_update` function.

zkat2 replied:
I went ahead and reintroduced a finish function to allow for multiple consecutive calls to update, indicate the calculation is complete, and release the semaphore.

Commit:
samples: drivers: crc Add sample for hardware based CRC API

Introduces a new sample that demonstrates how to use the
hardware-based CRC API.

Signed-off-by: Zoe Kaute <zoe.kaute@brillpower.com>

Commit:
tests: drivers: crc: Add test suite for hardware-based CRC API

Add a test suite to validate the hardware-based CRC API.

Signed-off-by: Zoe Kaute <zoe.kaute@brillpower.com>

Commit:
dts: arm: st: g4: Add CRC node

Add the CRC node to stm32g4 device tree source file.

Signed-off-by: Zoe Kaute <zoe.kaute@brillpower.com>

pdgendt commented on drivers/crc/crc_stm32.c:
```diff
@@ -0,0 +1,221 @@
+static int crc_stm32_begin(const struct device *dev, struct crc_ctx *ctx)
+{
+	/* Ensure ctx is not currently being updated */
+	if (ctx->state == CRC_STATE_IN_PROGRESS) {
+		return -EBUSY;
+	}
```
I think you can have a race issue here, checking the state outside of the lock, but changing it inside the lock.

erwango commented on samples/drivers/crc/src/main.c:
```diff
@@ -0,0 +1,71 @@
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+/* Sample usage for CRC_STM32 driver */
```
Nit: Samples should be vendor agnostic (and this is actually the case already)

erwango commented on tests/drivers/crc/src/main.c:
```diff
@@ -0,0 +1,225 @@
+				K_THREAD_STACK_SIZEOF(wait_thread_stack_area), wait_thread_entry,
+				NULL, NULL, NULL, WAIT_THREAD_PRIO, 0, K_NO_WAIT);
+
+	UNUSED(wait_tid);
```
`UNUSED` is not a zephyr API. Actually, you don't need to create `wait_tid`.

erwango commented on dts/arm/st/g4/stm32g474.dtsi:
```diff
@@ -7,7 +7,19 @@
+			reg = <0x40023000 0xA0>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x1000>;
+
+			status = "okay";
```
Devices should be enabled only in boards.
So please move status activation in matching nucleo board.
Also please mention `crc` support in board's yaml file. Then, in test and sample yaml files add `depends_on: crc`.
Then, you can remove `platform_(allow|exclude)` fields.

erwango commented on tests/drivers/crc/src/main.c:
```diff
@@ -0,0 +1,225 @@
+ */
+static void wait_thread_entry(void *a, void *b, void *c)
+{
+	UNUSED(a);
+	UNUSED(b);
+	UNUSED(c);
```
s/UNUSED|ARG_UNUSED

IIRC UNUSED is provided by STM32Cube HAL and should not be used in Zephyr

erwango commented on dts/arm/st/g4/stm32g474.dtsi:
```diff
@@ -7,7 +7,19 @@
 
 / {
+	aliases {
+		crc = &crc;
```
Aliases are usually called with `0` suffix. Ie `crc0` in this case.
I agree we should not see multiple crc instances, but this is a convention used (same for watchdog, lora, ...)

erwango commented on dts/arm/st/g4/stm32g474.dtsi:
```diff
@@ -7,7 +7,19 @@
 	soc {
 		compatible = "st,stm32g474", "st,stm32g4", "simple-bus";
+
+		crc: crc@40023000 {
```
crc is available in all G4 Socs, node should be added in g4.dtsi

erwango reviewed (changes requested):
Few more comments but we're getting there, thanks for the work so far.

Also, as you're introducing a new API, I've added flag "Architecture review", so it could be introduced to Architecture working group on next session (Tuesday 5PM CET, see https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group). It would be nice that you're able to join, but it is not blocking if you can't .

zkat2 commented on drivers/crc/Kconfig.stm32:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2024 Brill Power Ltd.
+# SPDX-License-Identifier: Apache-2.0
+
+config CRC_STM32
+	bool "STM32 CRC driver"
```
Removing this string causes a CMake error, are you provide an example of how to correctly make this non-user selectable?

erwango replied:
So indeed, I had a try and I'm facing the same issue.
Problem is that you're setting `CONFIG_CRC_STM32=y` in test and sample prj.conf files, which you don't need (as set depending on conditions defined here) and you shouldn't (as tests and samples should remain vendor agnostic).
Removing `CONFIG_CRC_STM32=y` from test and sample prj.conf files will solve the cmake issue and make the PR even more coherent.

henrikbrixandersen commented:
Any plans to make this driver subsystem provide backends for https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/crc.h ?

zkat2 commented on MAINTAINERS.yml:
```diff
@@ -1100,6 +1100,22 @@ Release Notes:
+"Drivers: CRC":
+  status: maintained
+  maintainers:
+    - zkat2
```
https://github.com/zephyrproject-rtos/zephyr/issues/76722

carlescufi commented:
Architecture WG:

- @gmarull and @carlescufi object with the fact that we are now left with 2 completely separate CRC APIs: the one in this PR and the existing software-based one in `sys/crc.h`
- One option would be to introduce something akin to the random subsystem, where there is a single API and then multiple backends, one of which uses the entropy driver API
- The other option would be to keep the API in this PR but add a software implementation, but then instancing may be an issue

ceolin commented:
> Architecture WG:
> 
> * @gmarull and @carlescufi object with the fact that we are now left with 2 completely separate CRC APIs: the one in this PR and the existing software-based one in `sys/crc.h`
> * One option would be to introduce something akin to the random subsystem, where there is a single API and then multiple backends, one of which uses the entropy driver API
> * The other option would be to keep the API in this PR but add a software implementation, but then instancing may be an issue
> * Check if mbedTLS has support for CRC APIs? @ceolin

As far as I can tell you there is no CRC.

> * @zkat2 feel free to join the Architecture WG to discuss this again

henrikbrixandersen commented:
@zkat2 Hey! Sorry to see this work abandoned. Do you have any plans - and bandwidth - for resuming work on this? I'd love to collaborate on it, if needed.

thenguyenyf commented:
> @zkat2 Hey! Sorry to see this work abandoned. Do you have any plans - and bandwidth - for resuming work on this? I'd love to collaborate on it, if needed.

Hello @henrikbrixandersen . We are preparing for a CRC hardware-based device driver (which similar to implementation in this PR) and we also introduce a CRC subsystem (implementation is quite similar to random system) includes a software backend (using sys/crc.h) and a CRC hardware backend (using CRC device driver). It will be sent out at the end of this week.

thenguyenyf commented:
Hello. I have sent out a PR to add a CRC device driver implementation for Renesas RA device that referred to this PR here https://github.com/zephyrproject-rtos/zephyr/pull/87557. I'm also implement a new CRC subsystem using software/hardware backend implementation.
@henrikbrixandersen , @zkat2 , could you please visit and leave it some comment?
